src/Microsoft.OData.Core/SelectedPropertiesNode.cs

```diff
@@ -469,6 +581,74 @@ private static IEnumerable<string> GetPossibleMatchesForSelectedOperation(IEdmOp
             yield return qualifiedContainerName + operationNameWithParameters;
         }
 
+        private bool IsValidExpandToken(string item)
+        {
+            // Check #1: Expand token must contain '(' and end with ')', must be related to navigation property.
+            int idxLP = item.IndexOf('(');
+            if (idxLP == -1
+                || !item.EndsWith(")", StringComparison.Ordinal)
+                || (this.structuredType != null && IsBoundOperationToken(item.Substring(0, idxLP))))
+            {
+                // selected item is not properly parenthesized, or is an operation token.
+                return false;
+            }
+
+            // Check #2: parentheses must be balanced.
+            int count = 0;
+            foreach (char c in item)
+            {
+                if (c == '(')
+                {
+                    ++count;
+                }
+                else if (c == ')')
+                {
+                    --count;
+                }
+
+                if (count < 0)
+                {
+                    // Malformed parentheses detected ==> not expand token.
+                    return false;
+                }
+            }
+
+            return count == 0;
+        }
+
+        /// <summary>
+        /// Checks the specified model and structuredType and see whether the token can be resolved to a bound operation name.
+        /// </summary>
+        /// <param name="token">The token to check.</param>
+        /// <returns>true if token can be matched a bound operation; false otherwise.</returns>
+        private bool IsBoundOperationToken(string token)
+        {
+            const char nameSpaceSeparator = '.';
+
+            if (this.edmModel == null)
+            {
+                return false;
+            }
+
+            bool found = false;
+            int idx = token.LastIndexOf(nameSpaceSeparator);
+            if (idx == -1)
+            {
+                // Token is in unqualified form, try matching the unqualified name of the bound operation.

```

Comment:
>Token is in unqualified form, try matching the unqualified name of the bound operation. [](start = 19, length = 87)

Should probably add comment that, according to spec, the name should always be qualified, but that older versions of the library returned unqualified.  And, we should fix the library such that it no longer returns unqualified names in the contextUrl.

Reply:
Added.

---
In reply to: [224139113](https://github.com/OData/odata.net/pull/1286#discussion_r224139113) [](ancestors = 224139113)